*** /tmp/,RCSt1023661	Wed Aug 29 14:31:44 1990
--- set_key.c	Wed Aug 29 14:29:43 1990
***************
*** 6,17 ****
   */
  #include "des_local.h"
  #include "podd.h"
  #include "bits.h"
  #include "ksp.h"
  
  static void norm_expand_bits();
  static int check_parity();
! static int make_key_sched();
  
  int des_check_key=0;
  
--- 6,19 ----
   */
  #include "des_local.h"
  #include "podd.h"
+ #ifdef notdef
  #include "bits.h"
  #include "ksp.h"
  
  static void norm_expand_bits();
+ #endif
  static int check_parity();
! extern int make_key_sched();
  
  int des_check_key=0;
  
***************
*** 94,100 ****
  des_cblock *key;
  des_key_schedule schedule;
  	{
- 	char k[64];
  
  	if (des_check_key)
  		{
--- 96,101 ----
***************
*** 105,115 ****
  			return(-2);
  		}
  
! 	norm_expand_bits(key,k,64);
! 	make_key_sched(k,schedule);
  	return(0);
  	}
  
  static void norm_expand_bits(from,to,num)
  register unsigned char *from;
  unsigned char *to;
--- 106,116 ----
  			return(-2);
  		}
  
! 	make_key_sched(key,schedule);
  	return(0);
  	}
  
+ #ifdef notdef
  static void norm_expand_bits(from,to,num)
  register unsigned char *from;
  unsigned char *to;
***************
*** 210,215 ****
--- 211,217 ----
  		}
  	return(0);
  	}
+ #endif
  
  int key_sched(key,schedule)
  des_cblock *key;
