*** kernel/ksyms.c.orig	Sat Jul  2 16:40:53 1994
--- kernel/ksyms.c	Sat Jul  2 16:43:23 1994
***************
*** 63,68 ****
--- 63,71 ----
  extern void dev_kfree_skb(struct sk_buff *, int);
  #endif
  
+ /* Some defines for the block driver */
+ #include "../drivers/block/blk.h"
+  
  struct symbol_table symbol_table = { 0, 0, 0, /* for stacked module support */
  	{
  	/* stackable module support */
***************
*** 178,183 ****
--- 182,201 ----
  	X(dev_kfree_skb),
  #endif
  
+ 	X(blk_dev),
+ 	X(blksize_size),
+ 	X(need_resched),
+ 	X(wait_for_request),
+ 	X(blk_size),
+ 	X(bmap),
+ 	X(getblk),
+ 	X(ll_rw_block),
+ 	X(__wait_on_buffer),
+ 	X(brelse),
+ 	X(invalidate_buffers),
+ 	X(block_read),
+ 	X(block_write),
+ 	X(sync_dev),
+ 
  	/********************************************************
  	 * Do not add anything below this line,
  	 * as the stacked modules depend on this!
