*** rpemmain.c~	Tue Mar 24 12:17:33 1992
--- rpemmain.c	Mon Sep 14 16:31:43 1992
***************
*** 1366,1371 ****
--- 1366,1374 ----
      * exchanged with buf2 (flushing buf2 to output if necessary).
      */
     while(fgets(line,LINEBUFSIZE,instream)) {
+ 	  if(strncmp(line,HEADER_STRING,strlen(HEADER_STRING))==0) 
+ 		break;
+ 
        bytes_in_line = prdecode(line,bufptr,LINEBUFSIZE);
        /* Compute the number of bytes that we won't be able to
         * decrypt in this pass.
