diff -c pgp263/src/fileio.c pgp263-beos/src/fileio.c
*** pgp263/src/fileio.c	Wed Apr 24 16:53:56 1996
--- pgp263-beos/src/fileio.c	Fri Feb 28 03:24:00 1997
***************
*** 377,382 ****
--- 377,384 ----
  #ifndef MAX_NAMELEN
  #if defined(AMIGA) || defined(NeXT) || (defined(BSD) && BSD > 41) || (defined(sun) && defined(i386))
  #define	MAX_NAMELEN	255
+ #elif defined(__BEOS__)
+ #define MAX_NAMELEN 63
  #else
  #ifdef MACTC5
  #define MAX_NAMELEN 31
***************
*** 1635,1640 ****
--- 1637,1646 ----
      "$PGPPATH", "", "[pgp]", "[pgp26]", "[pgp263]",
      PGP_SYSTEM_DIR, "SYS$LOGIN:", "SYS$LOGIN:[pgp]",
      "SYS$LOGIN:[pgp26]", "SYS$LOGIN:[pgp263]", "[-]",
+ #elif defined(__BEOS__)
+ 	"$PGPPATH", "", "pgp", "pgp26", "pgp263", "$HOME/.pgp",
+ 	"/boot/pgp", "/boot/local/pgp", "/boot/pgp263", 
+ 	"/boot/local/pgp263", "..",
  #elif defined(UNIX)
      "$PGPPATH", "", "pgp", "pgp26", "pgp263", PGP_SYSTEM_DIR,
      "$HOME/.pgp", "$HOME", "$HOME/pgp", "$HOME/pgp26", "..",
diff -c pgp263/src/fileio.h pgp263-beos/src/fileio.h
*** pgp263/src/fileio.h	Sat Jan  6 19:46:36 1996
--- pgp263-beos/src/fileio.h	Fri Feb 28 03:15:42 1997
***************
*** 27,33 ****
   */
  #ifdef LINUX
  #  define PGP_SYSTEM_DIR "/var/lib/pgp/"
! #else
  #  define PGP_SYSTEM_DIR "/usr/local/lib/pgp/"
  #endif
  #define FOPRBIN		"r"
--- 27,33 ----
   */
  #ifdef LINUX
  #  define PGP_SYSTEM_DIR "/var/lib/pgp/"
! #elif !defined(__BEOS__) /* Don't use this for BeOS 1.1d8 */
  #  define PGP_SYSTEM_DIR "/usr/local/lib/pgp/"
  #endif
  #define FOPRBIN		"r"
diff -c pgp263/src/getopt.c pgp263-beos/src/getopt.c
*** pgp263/src/getopt.c	Mon Jun 26 18:12:14 1995
--- pgp263-beos/src/getopt.c	Fri Feb 28 02:07:16 1997
***************
*** 26,32 ****
--- 26,34 ----
  #ifndef NULL
  #define NULL	0
  #endif
+ #ifndef EOF
  #define EOF	(-1)
+ #endif
  #define ERR(str, chr) (opterr ? \
  fprintf(stderr, "%s%s%c\n", argv[0], str, chr) : 0)
  
diff -c pgp263/src/makefile pgp263-beos/src/makefile
*** pgp263/src/makefile	Thu Jan 18 10:42:36 1996
--- pgp263-beos/src/makefile	Fri Feb 28 02:16:48 1997
***************
*** 78,84 ****
  	@echo ""
  	@echo "where <system> can be:"
  	@echo "	386bsd, 3b1, 3b1_asm, aix370, aix386, amix-68k-gcc, apollo,"
! 	@echo "	apollo-gcc, atari,aux(*), aux-gcc, bsd, bsdgcc, convexos,"
  	@echo "	dgux, djgpp, encore, freebsd, hpux-68k(*), hpux-68k-ansi,"
  	@echo "	hpux-68k-gcc, hpux-pa(*), hpux-pa-ansi, hpux-pa-gcc,"
  	@echo "	hpux9-pa-ansi, irix, irix_asm, isc, isc_asm, linux, linux-68k,"
--- 78,84 ----
  	@echo ""
  	@echo "where <system> can be:"
  	@echo "	386bsd, 3b1, 3b1_asm, aix370, aix386, amix-68k-gcc, apollo,"
! 	@echo "	apollo-gcc, atari,aux(*), aux-gcc, beos, bsd, bsdgcc, convexos,"
  	@echo "	dgux, djgpp, encore, freebsd, hpux-68k(*), hpux-68k-ansi,"
  	@echo "	hpux-68k-gcc, hpux-pa(*), hpux-pa-ansi, hpux-pa-gcc,"
  	@echo "	hpux9-pa-ansi, irix, irix_asm, isc, isc_asm, linux, linux-68k,"
***************
*** 574,579 ****
--- 574,585 ----
  	LD=gcc LIBS_EXT=-lpc RM=del
  	strip pgp
  	coff2exe pgp
+ 
+ # BeOS 1.1d8.2
+ beos:
+ 	$(MAKE) all CC=mwcc LD=mwcc \
+ 	CFLAGS="$(RSAINCDIR) -O7 -DUNIX -DHIGHFIRST -DIDEA32 \
+ 	-DPORTABLE -DMERRITT" LDFLAGS="-nodup"
  
  #
  # unproto for K&R compilers
diff -c pgp263/src/system.c pgp263-beos/src/system.c
*** pgp263/src/system.c	Sat Jan  6 20:02:55 1996
--- pgp263-beos/src/system.c	Fri Feb 28 01:05:51 1997
***************
*** 121,130 ****
--- 121,134 ----
  void ttycbreak(void)
  {
  	if (ttyfd == -1) {
+ #ifndef __BEOS__
  		if ((ttyfd = open("/dev/tty", O_RDWR)) < 0) {
  		    fprintf(stderr, "cannot open tty, using stdin\n");
  			ttyfd = 0;
  		}
+ #else /* BeOS 1.1d8 /dev/tty is broken */
+ 		ttyfd = 0;
+ #endif /* not __BEOS__ */
  	}
  #ifndef NOTERMIO
  #ifndef SVR2
