*** ../../gmp.h	Thu Jul 15 17:35:24 1993
--- gmp.h	Wed Aug 11 12:14:36 1993
***************
*** 283,289 ****
  mp_size_t mpn_bingcd (mp_ptr, mp_ptr, mp_size_t, mp_ptr, mp_size_t);
  mp_size_t mpn_accelgcd (mp_ptr, mp_ptr, mp_size_t, mp_ptr, mp_size_t);
  mp_size_t mpn_bmod (mp_ptr, mp_size_t, mp_srcptr, mp_size_t);
- mp_size_t mpn_bmodgcd (mp_ptr, mp_ptr, mp_size_t, mp_ptr, mp_size_t);
  void mpn_compl (mp_ptr, mp_srcptr, mp_size_t);
  mp_limb mpn_compl_rsh (mp_ptr, mp_srcptr, mp_size_t, unsigned int);
  mp_limb mpn_sub_n_rsh (mp_ptr, mp_srcptr, mp_srcptr, mp_size_t, unsigned int);
--- 283,288 ----
