*** /usr/src/sys/kern/uipc_socket2.c	Fri Feb  3 02:55:24 1995
--- uipc_socket2.c	Thu Sep 28 14:59:29 1995
***************
*** 1,3 ****
--- 1,10 ----
+ /*
+  * This file has been slightly modified by NRL for use with IPv6.  Search
+  * for 'INET6' to see the ifdef blocks where this happens.
+  *
+  * See the NRL Copyright notice for conditions on the modifications.
+  */
+ 
  /*	BSDI uipc_socket2.c,v 2.1 1995/02/03 07:55:24 polk Exp	*/
  
  /*
***************
*** 167,172 ****
--- 174,193 ----
  	so->so_proto = head->so_proto;
  	so->so_timeo = head->so_timeo;
  	so->so_pgid = head->so_pgid;
+ #ifdef INET6
+ 
+ 	/*
+ 	 * Copy over security levels to child socket.
+ 	 * GAG!  How do I determine that last argument to bcopy?
+ 	 */
+ 
+ 	bcopy(head->so_securitylevel, so->so_securitylevel, 4 * sizeof(int));
+ 
+ 	/*
+ 	 * Kebe says you might have to work some magic with the key mgmt.
+ 	 * infrastructure, too.  (But it's looking like I won't have to!!! :)
+ 	 */
+ #endif /* INET6 */
  	(void) soreserve(so, head->so_snd.sb_hiwat, head->so_rcv.sb_hiwat);
  	soqinsque(head, so, soqueue);
  	if ((*so->so_proto->pr_usrreq)(so, PRU_ATTACH,
