Index: clients/pgp/cmdline/globals.h
===================================================================
RCS file: /pgpsrc/clients/pgp/cmdline/globals.h,v
retrieving revision 1.18.6.2
retrieving revision 1.18.2.1.2.1
diff -u -r1.18.6.2 -r1.18.2.1.2.1
--- globals.h	1999/06/14 19:01:57	1.18.6.2
+++ globals.h	1999/07/19 17:41:48	1.18.2.1.2.1
@@ -8,7 +8,7 @@
 
 	struct wrappers around the former pgp262 global variables
 
-    $Id: globals.h,v 1.18.6.2 1999/06/14 19:01:57 build Exp $
+    $Id: globals.h,v 1.18.2.1.2.1 1999/07/19 17:41:48 sluu Exp $
 ____________________________________________________________________________*/
 
 #ifndef PGP_CMDLINE_GLOBALS
@@ -18,6 +18,7 @@
 #include "pgpDebug.h"
 #include "pgpEnv.h"
 #include "pgpGroups.h"
+#include "pgpDebug.h"
 
 #include "lists.h"
 
Index: libs/network/ldaplib/Makefile
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/Makefile,v
retrieving revision 1.4
retrieving revision 1.4.2.1
diff -u -r1.4 -r1.4.2.1
--- Makefile	1999/03/25 17:55:12	1.4
+++ Makefile	1999/06/10 06:08:15	1.4.2.1
@@ -237,7 +237,7 @@
 	CC=$(CC); \
 	OLDIFS="$$IFS"; \
 	IFS=":"; \
-if test ! -n "$CC"; then \
+if test -n "$CC"; then \
 	for dir in $$PATH; do \
 	    if [ -f $$dir/gcc ]; then \
 		CC=gcc; \
Index: libs/network/ldaplib/build/Makefile
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/build/Makefile,v
retrieving revision 1.6.10.1
retrieving revision 1.6.2.1
diff -u -r1.6.10.1 -r1.6.2.1
--- Makefile	1999/07/01 01:47:37	1.6.10.1
+++ Makefile	1999/06/18 19:31:59	1.6.2.1
@@ -122,33 +122,24 @@
 FORCE:
 
 #
-# LDAP SunOS5 GNU C Make-platform file
+# LDAP AIX standard cc Make-platform file
 #
 
+
 #
 # add any platform-specific overrides below here
 #
 
-
 #
 # -------------------------------------------------------------------------
 # you will probably not need to edit anything below this point
 # -------------------------------------------------------------------------
 CC	= gcc
-
-# ranlib not needed under SunOS5
-RANLIB	= ""
-
-# give full path to hostname since it may not be in user's path
-HOSTNAME=/usr/ucb/hostname
 
-# don't count on /usr/ucb/install being present or first in path
-INSTALL=$(LDAPSRC)/build/install.sh
+# install with BSD commandline
+INSTALL=/usr/ucb/install
 
-PLATFORMCFLAGS=	-Dsunos5 -D_REENTRANT
-PLATFORMLIBS= -lsocket -lnsl -lgen
-THREADS= -DTHREAD_SUNOS5_LWP
-THREADSLIB=-lthread
+PLATFORMCFLAGS=	-Daix
 #-----------------------------------------------------------------------------
 # Copyright (c) 1997 PGP inc.
 # All rights reserved.
Index: libs/network/ldaplib/include/Makefile
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/include/Makefile,v
retrieving revision 1.6.10.1
retrieving revision 1.6.2.1
diff -u -r1.6.10.1 -r1.6.2.1
--- Makefile	1999/07/01 01:47:37	1.6.10.1
+++ Makefile	1999/06/18 19:31:59	1.6.2.1
@@ -147,33 +147,24 @@
 FORCE:
 
 #
-# LDAP SunOS5 GNU C Make-platform file
+# LDAP AIX standard cc Make-platform file
 #
 
+
 #
 # add any platform-specific overrides below here
 #
 
-
 #
 # -------------------------------------------------------------------------
 # you will probably not need to edit anything below this point
 # -------------------------------------------------------------------------
 CC	= gcc
-
-# ranlib not needed under SunOS5
-RANLIB	= ""
-
-# give full path to hostname since it may not be in user's path
-HOSTNAME=/usr/ucb/hostname
 
-# don't count on /usr/ucb/install being present or first in path
-INSTALL=$(LDAPSRC)/build/install.sh
+# install with BSD commandline
+INSTALL=/usr/ucb/install
 
-PLATFORMCFLAGS=	-Dsunos5 -D_REENTRANT
-PLATFORMLIBS= -lsocket -lnsl -lgen
-THREADS= -DTHREAD_SUNOS5_LWP
-THREADSLIB=-lthread
+PLATFORMCFLAGS=	-Daix
 #-----------------------------------------------------------------------------
 # Copyright (c) 1997 PGP inc.
 # All rights reserved.
Index: libs/network/ldaplib/include/ldap.h
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/include/ldap.h,v
retrieving revision 1.10
retrieving revision 1.10.4.2.2.1
diff -u -r1.10 -r1.10.4.2.2.1
--- ldap.h	1999/02/25 21:15:08	1.10
+++ ldap.h	1999/07/19 17:41:50	1.10.4.2.2.1
@@ -86,7 +86,8 @@
 #define GetThreadID		GetCurrentThreadId
 #elif PGP_UNIX_SOLARIS
 #define GetThreadID		thr_self
-#elif PGP_UNIX_LINUX
+#elif PGP_UNIX_LINUX || PGP_UNIX_AIX || PGP_UNIX_HPUX
+/* This is for all pthreads systems, including Linux and AIX. */
 #define GetThreadID		pthread_self
 #endif
 
Index: libs/network/ldaplib/liblber/Make-template
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/liblber/Make-template,v
retrieving revision 1.10.10.1
retrieving revision 1.10.2.2
diff -u -r1.10.10.1 -r1.10.2.2
--- Make-template	1999/07/01 01:47:37	1.10.10.1
+++ Make-template	1999/06/24 23:12:32	1.10.2.2
@@ -89,31 +89,26 @@
 decode.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 decode.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 decode.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-decode.o: ../../../pgpcdk/pub/include/pgpTLS.h
 decode.o: ../../../pgpcdk/pub/include/pgpErrors.h
 decode.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 encode.o: encode.c ../pgpsockets/pgpSocketsLDAP.h ../include/lber.h
 encode.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 encode.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 encode.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-encode.o: ../../../pgpcdk/pub/include/pgpTLS.h
 encode.o: ../../../pgpcdk/pub/include/pgpErrors.h
 encode.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-io.o: io.c ../pgpsockets/pgpSocketsLDAP.h ../include/lber.h
-io.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+io.o: io.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 io.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 io.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-io.o: ../../../pgpcdk/pub/include/pgpTLS.h
 io.o: ../../../pgpcdk/pub/include/pgpErrors.h
 io.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 bprint.o: bprint.c ../../../pgpcdk/pub/include/pgpSockets.h
 bprint.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 bprint.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 bprint.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-bprint.o: ../../../pgpcdk/pub/include/pgpTLS.h
 bprint.o: ../../../pgpcdk/pub/include/pgpErrors.h
 bprint.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-pgpSrvMem.o: pgpSrvMem.c ../../../pfl/common/pgpBase.h
-pgpSrvMem.o: ../../../pfl/common/pgpMemoryMgr.h
+pgpSrvMem.o: pgpSrvMem.c ../../../pfl/common/pgpPFLErrors.h
+pgpSrvMem.o: ../../../pfl/unix/pgpPFLConfig.h ../include/pgpSrvStd.h
 
 # IF YOU PUT ANYTHING HERE IT WILL GO AWAY
Index: libs/network/ldaplib/liblber/Makefile
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/liblber/Makefile,v
retrieving revision 1.10.10.1
retrieving revision 1.10.2.2
diff -u -r1.10.10.1 -r1.10.2.2
--- Makefile	1999/07/01 01:47:37	1.10.10.1
+++ Makefile	1999/06/24 23:12:32	1.10.2.2
@@ -96,32 +96,27 @@
 decode.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 decode.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 decode.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-decode.o: ../../../pgpcdk/pub/include/pgpTLS.h
 decode.o: ../../../pgpcdk/pub/include/pgpErrors.h
 decode.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 encode.o: encode.c ../pgpsockets/pgpSocketsLDAP.h ../include/lber.h
 encode.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 encode.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 encode.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-encode.o: ../../../pgpcdk/pub/include/pgpTLS.h
 encode.o: ../../../pgpcdk/pub/include/pgpErrors.h
 encode.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-io.o: io.c ../pgpsockets/pgpSocketsLDAP.h ../include/lber.h
-io.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+io.o: io.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 io.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 io.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-io.o: ../../../pgpcdk/pub/include/pgpTLS.h
 io.o: ../../../pgpcdk/pub/include/pgpErrors.h
 io.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 bprint.o: bprint.c ../../../pgpcdk/pub/include/pgpSockets.h
 bprint.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 bprint.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 bprint.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-bprint.o: ../../../pgpcdk/pub/include/pgpTLS.h
 bprint.o: ../../../pgpcdk/pub/include/pgpErrors.h
 bprint.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-pgpSrvMem.o: pgpSrvMem.c ../../../pfl/common/pgpBase.h
-pgpSrvMem.o: ../../../pfl/common/pgpMemoryMgr.h
+pgpSrvMem.o: pgpSrvMem.c ../../../pfl/common/pgpPFLErrors.h
+pgpSrvMem.o: ../../../pfl/unix/pgpPFLConfig.h ../include/pgpSrvStd.h
 
 # IF YOU PUT ANYTHING HERE IT WILL GO AWAY
 #
@@ -203,33 +198,24 @@
 FORCE:
 
 #
-# LDAP SunOS5 GNU C Make-platform file
+# LDAP AIX standard cc Make-platform file
 #
 
+
 #
 # add any platform-specific overrides below here
 #
 
-
 #
 # -------------------------------------------------------------------------
 # you will probably not need to edit anything below this point
 # -------------------------------------------------------------------------
 CC	= gcc
-
-# ranlib not needed under SunOS5
-RANLIB	= ""
-
-# give full path to hostname since it may not be in user's path
-HOSTNAME=/usr/ucb/hostname
 
-# don't count on /usr/ucb/install being present or first in path
-INSTALL=$(LDAPSRC)/build/install.sh
+# install with BSD commandline
+INSTALL=/usr/ucb/install
 
-PLATFORMCFLAGS=	-Dsunos5 -D_REENTRANT
-PLATFORMLIBS= -lsocket -lnsl -lgen
-THREADS= -DTHREAD_SUNOS5_LWP
-THREADSLIB=-lthread
+PLATFORMCFLAGS=	-Daix
 #-----------------------------------------------------------------------------
 # Copyright (c) 1997 PGP inc.
 # All rights reserved.
Index: libs/network/ldaplib/libldap/Make-template
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/libldap/Make-template,v
retrieving revision 1.7.10.1
retrieving revision 1.7.2.2
diff -u -r1.7.10.1 -r1.7.2.2
--- Make-template	1999/07/01 01:47:37	1.7.10.1
+++ Make-template	1999/06/24 23:12:32	1.7.2.2
@@ -112,7 +112,6 @@
 bind.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 bind.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 bind.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-bind.o: ../../../pgpcdk/pub/include/pgpTLS.h
 bind.o: ../../../pgpcdk/pub/include/pgpErrors.h
 bind.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 bind.o: ../include/ldap.h
@@ -121,15 +120,14 @@
 open.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 open.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 open.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-open.o: ../../../pgpcdk/pub/include/pgpTLS.h
 open.o: ../../../pgpcdk/pub/include/pgpErrors.h
 open.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 open.o: ../include/ldap.h
-result.o: result.c ../include/lber.h
+result.o: result.c ../pgpsockets/pgpSocketsLDAP.h
+result.o: ../../../pgpcdk/pub/include/pgpSockets.h
 result.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 result.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 result.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-result.o: ../../../pgpcdk/pub/include/pgpTLS.h
 result.o: ../../../pgpcdk/pub/include/pgpErrors.h
 result.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 result.o: ../include/ldap.h
@@ -138,32 +136,29 @@
 error.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 error.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 error.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-error.o: ../../../pgpcdk/pub/include/pgpTLS.h
 error.o: ../../../pgpcdk/pub/include/pgpErrors.h
 error.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 error.o: ../include/ldap.h
-compare.o: compare.c ../include/lber.h
+compare.o: compare.c ../pgpsockets/pgpSocketsLDAP.h
+compare.o: ../../../pgpcdk/pub/include/pgpSockets.h
 compare.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 compare.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 compare.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-compare.o: ../../../pgpcdk/pub/include/pgpTLS.h
 compare.o: ../../../pgpcdk/pub/include/pgpErrors.h
 compare.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 compare.o: ../include/ldap.h
-search.o: search.c ../pgpsockets/pgpSocketsLDAP.h
-search.o: ../../../pgpcdk/pub/include/pgpSockets.h
+search.o: search.c ../include/lber.h
 search.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 search.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 search.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-search.o: ../../../pgpcdk/pub/include/pgpTLS.h
 search.o: ../../../pgpcdk/pub/include/pgpErrors.h
 search.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-search.o: ../include/ldap.h
-modify.o: modify.c ../include/lber.h
+search.o: ../include/proto-ldap.h
+modify.o: modify.c ../pgpsockets/pgpSocketsLDAP.h
+modify.o: ../../../pgpcdk/pub/include/pgpSockets.h
 modify.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 modify.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 modify.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-modify.o: ../../../pgpcdk/pub/include/pgpTLS.h
 modify.o: ../../../pgpcdk/pub/include/pgpErrors.h
 modify.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 modify.o: ../include/ldap.h
@@ -172,71 +167,64 @@
 add.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 add.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 add.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-add.o: ../../../pgpcdk/pub/include/pgpTLS.h
 add.o: ../../../pgpcdk/pub/include/pgpErrors.h
 add.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 add.o: ../include/ldap.h
-modrdn.o: modrdn.c ../include/lber.h
+modrdn.o: modrdn.c ../pgpsockets/pgpSocketsLDAP.h
+modrdn.o: ../../../pgpcdk/pub/include/pgpSockets.h
 modrdn.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 modrdn.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 modrdn.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-modrdn.o: ../../../pgpcdk/pub/include/pgpTLS.h
 modrdn.o: ../../../pgpcdk/pub/include/pgpErrors.h
 modrdn.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 modrdn.o: ../include/ldap.h
-delete.o: delete.c ../include/lber.h
+delete.o: delete.c ../pgpsockets/pgpSocketsLDAP.h
+delete.o: ../../../pgpcdk/pub/include/pgpSockets.h
 delete.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 delete.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 delete.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-delete.o: ../../../pgpcdk/pub/include/pgpTLS.h
 delete.o: ../../../pgpcdk/pub/include/pgpErrors.h
 delete.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 delete.o: ../include/ldap.h
-abandon.o: abandon.c ../include/lber.h
+abandon.o: abandon.c ../pgpsockets/pgpSocketsLDAP.h
+abandon.o: ../../../pgpcdk/pub/include/pgpSockets.h
 abandon.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 abandon.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 abandon.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-abandon.o: ../../../pgpcdk/pub/include/pgpTLS.h
 abandon.o: ../../../pgpcdk/pub/include/pgpErrors.h
 abandon.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 abandon.o: ../include/ldap.h
-ufn.o: ufn.c ../pgpsockets/pgpSocketsLDAP.h
-ufn.o: ../../../pgpcdk/pub/include/pgpSockets.h
-ufn.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+ufn.o: ufn.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 ufn.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 ufn.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-ufn.o: ../../../pgpcdk/pub/include/pgpTLS.h
 ufn.o: ../../../pgpcdk/pub/include/pgpErrors.h
 ufn.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-ufn.o: ../include/ldap.h
+ufn.o: ../include/proto-ldap.h
 cache.o: cache.c ../../../pgpcdk/pub/include/pgpSockets.h
 cache.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 cache.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 cache.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-cache.o: ../../../pgpcdk/pub/include/pgpTLS.h
 cache.o: ../../../pgpcdk/pub/include/pgpErrors.h
-cache.o: ../../../pfl/common/pgpPFLErrors.h ../include/proto-lber.h ldap-int.h
-getfilter.o: getfilter.c ../pgpsockets/pgpSocketsLDAP.h
-getfilter.o: ../../../pgpcdk/pub/include/pgpSockets.h
+cache.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+cache.o: ../include/ldap.h
+getfilter.o: getfilter.c ../include/lber.h
 getfilter.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getfilter.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getfilter.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getfilter.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getfilter.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getfilter.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getfilter.o: ../include/ldap.h
+getfilter.o: ../include/proto-ldap.h
 regex.o: regex.c
 sbind.o: sbind.c ../pgpsockets/pgpSocketsLDAP.h
 sbind.o: ../../../pgpcdk/pub/include/pgpSockets.h
 sbind.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 sbind.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 sbind.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-sbind.o: ../../../pgpcdk/pub/include/pgpTLS.h
 sbind.o: ../../../pgpcdk/pub/include/pgpErrors.h
 sbind.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 sbind.o: ../include/ldap.h
 kbind.o: kbind.c
-unbind.o: unbind.c ../../../pgpcdk/pub/include/pgpTLS.h
+unbind.o: unbind.c ../pgpsockets/pgpSocketsLDAP.h
 unbind.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 unbind.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
 unbind.o: ../../../pgpcdk/pub/include/pgpSockets.h
@@ -249,54 +237,48 @@
 friendly.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 friendly.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 friendly.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-friendly.o: ../../../pgpcdk/pub/include/pgpTLS.h
 friendly.o: ../../../pgpcdk/pub/include/pgpErrors.h
 friendly.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-friendly.o: ../include/ldap.h
+friendly.o: ../include/proto-ldap.h
 cldap.o: cldap.c
-free.o: free.c ../pgpsockets/pgpSocketsLDAP.h
-free.o: ../../../pgpcdk/pub/include/pgpSockets.h
-free.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+free.o: free.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 free.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 free.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-free.o: ../../../pgpcdk/pub/include/pgpTLS.h
 free.o: ../../../pgpcdk/pub/include/pgpErrors.h
 free.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-free.o: ../include/ldap.h
+free.o: ../include/proto-ldap.h
 disptmpl.o: disptmpl.c ../../../pgpcdk/pub/include/pgpSockets.h
 disptmpl.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 disptmpl.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 disptmpl.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-disptmpl.o: ../../../pgpcdk/pub/include/pgpTLS.h
 disptmpl.o: ../../../pgpcdk/pub/include/pgpErrors.h
-disptmpl.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+disptmpl.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+disptmpl.o: ../include/proto-ldap.h
 srchpref.o: srchpref.c ../../../pgpcdk/pub/include/pgpSockets.h
 srchpref.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 srchpref.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 srchpref.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-srchpref.o: ../../../pgpcdk/pub/include/pgpTLS.h
 srchpref.o: ../../../pgpcdk/pub/include/pgpErrors.h
-srchpref.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+srchpref.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+srchpref.o: ../include/proto-ldap.h
 dsparse.o: dsparse.c ../pgpsockets/pgpSocketsLDAP.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpSockets.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 dsparse.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-dsparse.o: ../../../pgpcdk/pub/include/pgpTLS.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpErrors.h
 dsparse.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-dsparse.o: ../include/ldap.h
+dsparse.o: ../include/proto-ldap.h
 tmplout.o: tmplout.c ../../../pgpcdk/pub/include/pgpSockets.h
 tmplout.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 tmplout.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 tmplout.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-tmplout.o: ../../../pgpcdk/pub/include/pgpTLS.h
 tmplout.o: ../../../pgpcdk/pub/include/pgpErrors.h
-tmplout.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+tmplout.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+tmplout.o: ../include/proto-ldap.h
 sort.o: sort.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 sort.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 sort.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-sort.o: ../../../pgpcdk/pub/include/pgpTLS.h
 sort.o: ../../../pgpcdk/pub/include/pgpErrors.h
 sort.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 sort.o: ../include/proto-ldap.h
@@ -305,51 +287,46 @@
 getdn.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getdn.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getdn.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getdn.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getdn.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getdn.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getdn.o: ../include/ldap.h
+getdn.o: ../include/proto-ldap.h
 getentry.o: getentry.c ../pgpsockets/pgpSocketsLDAP.h
 getentry.o: ../../../pgpcdk/pub/include/pgpSockets.h
 getentry.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getentry.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getentry.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getentry.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getentry.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getentry.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getentry.o: ../include/ldap.h
+getentry.o: ../include/proto-ldap.h
 getattr.o: getattr.c ../pgpsockets/pgpSocketsLDAP.h
 getattr.o: ../../../pgpcdk/pub/include/pgpSockets.h
 getattr.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getattr.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getattr.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getattr.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getattr.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getattr.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getattr.o: ../include/ldap.h
+getattr.o: ../include/proto-ldap.h
 getvalues.o: getvalues.c ../pgpsockets/pgpSocketsLDAP.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpSockets.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getvalues.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getvalues.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getvalues.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getvalues.o: ../include/ldap.h
+getvalues.o: ../include/proto-ldap.h
 addentry.o: addentry.c ../pgpsockets/pgpSocketsLDAP.h
 addentry.o: ../../../pgpcdk/pub/include/pgpSockets.h
 addentry.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 addentry.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 addentry.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-addentry.o: ../../../pgpcdk/pub/include/pgpTLS.h
 addentry.o: ../../../pgpcdk/pub/include/pgpErrors.h
 addentry.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-addentry.o: ../include/ldap.h
-request.o: request.c ../include/lber.h
+addentry.o: ../include/proto-ldap.h
+request.o: request.c ../pgpsockets/pgpSocketsLDAP.h
+request.o: ../../../pgpcdk/pub/include/pgpSockets.h
 request.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 request.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 request.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-request.o: ../../../pgpcdk/pub/include/pgpTLS.h
 request.o: ../../../pgpcdk/pub/include/pgpErrors.h
 request.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 request.o: ../include/ldap.h
@@ -358,18 +335,14 @@
 os-ip.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 os-ip.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 os-ip.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-os-ip.o: ../../../pgpcdk/pub/include/pgpTLS.h
 os-ip.o: ../../../pgpcdk/pub/include/pgpErrors.h
-os-ip.o: ../../../pfl/common/pgpPFLErrors.h ../include/proto-ldap.h
-url.o: url.c ../pgpsockets/pgpSocketsLDAP.h
-url.o: ../../../pgpcdk/pub/include/pgpSockets.h
-url.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+os-ip.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+url.o: url.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 url.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 url.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-url.o: ../../../pgpcdk/pub/include/pgpTLS.h
 url.o: ../../../pgpcdk/pub/include/pgpErrors.h
 url.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-url.o: ../include/ldap.h
+url.o: ../include/proto-ldap.h
 charset.o: charset.c
 
 # IF YOU PUT ANYTHING HERE IT WILL GO AWAY
Index: libs/network/ldaplib/libldap/Makefile
===================================================================
RCS file: /pgpsrc/libs/network/ldaplib/libldap/Makefile,v
retrieving revision 1.8.10.1
retrieving revision 1.8.2.2
diff -u -r1.8.10.1 -r1.8.2.2
--- Makefile	1999/07/01 01:47:38	1.8.10.1
+++ Makefile	1999/06/24 23:12:33	1.8.2.2
@@ -119,7 +119,6 @@
 bind.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 bind.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 bind.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-bind.o: ../../../pgpcdk/pub/include/pgpTLS.h
 bind.o: ../../../pgpcdk/pub/include/pgpErrors.h
 bind.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 bind.o: ../include/ldap.h
@@ -128,15 +127,14 @@
 open.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 open.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 open.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-open.o: ../../../pgpcdk/pub/include/pgpTLS.h
 open.o: ../../../pgpcdk/pub/include/pgpErrors.h
 open.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 open.o: ../include/ldap.h
-result.o: result.c ../include/lber.h
+result.o: result.c ../pgpsockets/pgpSocketsLDAP.h
+result.o: ../../../pgpcdk/pub/include/pgpSockets.h
 result.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 result.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 result.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-result.o: ../../../pgpcdk/pub/include/pgpTLS.h
 result.o: ../../../pgpcdk/pub/include/pgpErrors.h
 result.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 result.o: ../include/ldap.h
@@ -145,32 +143,29 @@
 error.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 error.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 error.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-error.o: ../../../pgpcdk/pub/include/pgpTLS.h
 error.o: ../../../pgpcdk/pub/include/pgpErrors.h
 error.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 error.o: ../include/ldap.h
-compare.o: compare.c ../include/lber.h
+compare.o: compare.c ../pgpsockets/pgpSocketsLDAP.h
+compare.o: ../../../pgpcdk/pub/include/pgpSockets.h
 compare.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 compare.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 compare.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-compare.o: ../../../pgpcdk/pub/include/pgpTLS.h
 compare.o: ../../../pgpcdk/pub/include/pgpErrors.h
 compare.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 compare.o: ../include/ldap.h
-search.o: search.c ../pgpsockets/pgpSocketsLDAP.h
-search.o: ../../../pgpcdk/pub/include/pgpSockets.h
+search.o: search.c ../include/lber.h
 search.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 search.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 search.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-search.o: ../../../pgpcdk/pub/include/pgpTLS.h
 search.o: ../../../pgpcdk/pub/include/pgpErrors.h
 search.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-search.o: ../include/ldap.h
-modify.o: modify.c ../include/lber.h
+search.o: ../include/proto-ldap.h
+modify.o: modify.c ../pgpsockets/pgpSocketsLDAP.h
+modify.o: ../../../pgpcdk/pub/include/pgpSockets.h
 modify.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 modify.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 modify.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-modify.o: ../../../pgpcdk/pub/include/pgpTLS.h
 modify.o: ../../../pgpcdk/pub/include/pgpErrors.h
 modify.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 modify.o: ../include/ldap.h
@@ -179,71 +174,64 @@
 add.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 add.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 add.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-add.o: ../../../pgpcdk/pub/include/pgpTLS.h
 add.o: ../../../pgpcdk/pub/include/pgpErrors.h
 add.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 add.o: ../include/ldap.h
-modrdn.o: modrdn.c ../include/lber.h
+modrdn.o: modrdn.c ../pgpsockets/pgpSocketsLDAP.h
+modrdn.o: ../../../pgpcdk/pub/include/pgpSockets.h
 modrdn.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 modrdn.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 modrdn.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-modrdn.o: ../../../pgpcdk/pub/include/pgpTLS.h
 modrdn.o: ../../../pgpcdk/pub/include/pgpErrors.h
 modrdn.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 modrdn.o: ../include/ldap.h
-delete.o: delete.c ../include/lber.h
+delete.o: delete.c ../pgpsockets/pgpSocketsLDAP.h
+delete.o: ../../../pgpcdk/pub/include/pgpSockets.h
 delete.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 delete.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 delete.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-delete.o: ../../../pgpcdk/pub/include/pgpTLS.h
 delete.o: ../../../pgpcdk/pub/include/pgpErrors.h
 delete.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 delete.o: ../include/ldap.h
-abandon.o: abandon.c ../include/lber.h
+abandon.o: abandon.c ../pgpsockets/pgpSocketsLDAP.h
+abandon.o: ../../../pgpcdk/pub/include/pgpSockets.h
 abandon.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 abandon.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 abandon.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-abandon.o: ../../../pgpcdk/pub/include/pgpTLS.h
 abandon.o: ../../../pgpcdk/pub/include/pgpErrors.h
 abandon.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 abandon.o: ../include/ldap.h
-ufn.o: ufn.c ../pgpsockets/pgpSocketsLDAP.h
-ufn.o: ../../../pgpcdk/pub/include/pgpSockets.h
-ufn.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+ufn.o: ufn.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 ufn.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 ufn.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-ufn.o: ../../../pgpcdk/pub/include/pgpTLS.h
 ufn.o: ../../../pgpcdk/pub/include/pgpErrors.h
 ufn.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-ufn.o: ../include/ldap.h
+ufn.o: ../include/proto-ldap.h
 cache.o: cache.c ../../../pgpcdk/pub/include/pgpSockets.h
 cache.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 cache.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 cache.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-cache.o: ../../../pgpcdk/pub/include/pgpTLS.h
 cache.o: ../../../pgpcdk/pub/include/pgpErrors.h
-cache.o: ../../../pfl/common/pgpPFLErrors.h ../include/proto-lber.h ldap-int.h
-getfilter.o: getfilter.c ../pgpsockets/pgpSocketsLDAP.h
-getfilter.o: ../../../pgpcdk/pub/include/pgpSockets.h
+cache.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+cache.o: ../include/ldap.h
+getfilter.o: getfilter.c ../include/lber.h
 getfilter.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getfilter.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getfilter.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getfilter.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getfilter.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getfilter.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getfilter.o: ../include/ldap.h
+getfilter.o: ../include/proto-ldap.h
 regex.o: regex.c
 sbind.o: sbind.c ../pgpsockets/pgpSocketsLDAP.h
 sbind.o: ../../../pgpcdk/pub/include/pgpSockets.h
 sbind.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 sbind.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 sbind.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-sbind.o: ../../../pgpcdk/pub/include/pgpTLS.h
 sbind.o: ../../../pgpcdk/pub/include/pgpErrors.h
 sbind.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 sbind.o: ../include/ldap.h
 kbind.o: kbind.c
-unbind.o: unbind.c ../../../pgpcdk/pub/include/pgpTLS.h
+unbind.o: unbind.c ../pgpsockets/pgpSocketsLDAP.h
 unbind.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 unbind.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
 unbind.o: ../../../pgpcdk/pub/include/pgpSockets.h
@@ -256,54 +244,48 @@
 friendly.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 friendly.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 friendly.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-friendly.o: ../../../pgpcdk/pub/include/pgpTLS.h
 friendly.o: ../../../pgpcdk/pub/include/pgpErrors.h
 friendly.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-friendly.o: ../include/ldap.h
+friendly.o: ../include/proto-ldap.h
 cldap.o: cldap.c
-free.o: free.c ../pgpsockets/pgpSocketsLDAP.h
-free.o: ../../../pgpcdk/pub/include/pgpSockets.h
-free.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+free.o: free.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 free.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 free.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-free.o: ../../../pgpcdk/pub/include/pgpTLS.h
 free.o: ../../../pgpcdk/pub/include/pgpErrors.h
 free.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-free.o: ../include/ldap.h
+free.o: ../include/proto-ldap.h
 disptmpl.o: disptmpl.c ../../../pgpcdk/pub/include/pgpSockets.h
 disptmpl.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 disptmpl.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 disptmpl.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-disptmpl.o: ../../../pgpcdk/pub/include/pgpTLS.h
 disptmpl.o: ../../../pgpcdk/pub/include/pgpErrors.h
-disptmpl.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+disptmpl.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+disptmpl.o: ../include/proto-ldap.h
 srchpref.o: srchpref.c ../../../pgpcdk/pub/include/pgpSockets.h
 srchpref.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 srchpref.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 srchpref.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-srchpref.o: ../../../pgpcdk/pub/include/pgpTLS.h
 srchpref.o: ../../../pgpcdk/pub/include/pgpErrors.h
-srchpref.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+srchpref.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+srchpref.o: ../include/proto-ldap.h
 dsparse.o: dsparse.c ../pgpsockets/pgpSocketsLDAP.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpSockets.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 dsparse.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-dsparse.o: ../../../pgpcdk/pub/include/pgpTLS.h
 dsparse.o: ../../../pgpcdk/pub/include/pgpErrors.h
 dsparse.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-dsparse.o: ../include/ldap.h
+dsparse.o: ../include/proto-ldap.h
 tmplout.o: tmplout.c ../../../pgpcdk/pub/include/pgpSockets.h
 tmplout.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 tmplout.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 tmplout.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-tmplout.o: ../../../pgpcdk/pub/include/pgpTLS.h
 tmplout.o: ../../../pgpcdk/pub/include/pgpErrors.h
-tmplout.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+tmplout.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
+tmplout.o: ../include/proto-ldap.h
 sort.o: sort.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 sort.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 sort.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-sort.o: ../../../pgpcdk/pub/include/pgpTLS.h
 sort.o: ../../../pgpcdk/pub/include/pgpErrors.h
 sort.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 sort.o: ../include/proto-ldap.h
@@ -312,51 +294,46 @@
 getdn.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getdn.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getdn.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getdn.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getdn.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getdn.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getdn.o: ../include/ldap.h
+getdn.o: ../include/proto-ldap.h
 getentry.o: getentry.c ../pgpsockets/pgpSocketsLDAP.h
 getentry.o: ../../../pgpcdk/pub/include/pgpSockets.h
 getentry.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getentry.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getentry.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getentry.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getentry.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getentry.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getentry.o: ../include/ldap.h
+getentry.o: ../include/proto-ldap.h
 getattr.o: getattr.c ../pgpsockets/pgpSocketsLDAP.h
 getattr.o: ../../../pgpcdk/pub/include/pgpSockets.h
 getattr.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getattr.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getattr.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getattr.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getattr.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getattr.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getattr.o: ../include/ldap.h
+getattr.o: ../include/proto-ldap.h
 getvalues.o: getvalues.c ../pgpsockets/pgpSocketsLDAP.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpSockets.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 getvalues.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-getvalues.o: ../../../pgpcdk/pub/include/pgpTLS.h
 getvalues.o: ../../../pgpcdk/pub/include/pgpErrors.h
 getvalues.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-getvalues.o: ../include/ldap.h
+getvalues.o: ../include/proto-ldap.h
 addentry.o: addentry.c ../pgpsockets/pgpSocketsLDAP.h
 addentry.o: ../../../pgpcdk/pub/include/pgpSockets.h
 addentry.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 addentry.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 addentry.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-addentry.o: ../../../pgpcdk/pub/include/pgpTLS.h
 addentry.o: ../../../pgpcdk/pub/include/pgpErrors.h
 addentry.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-addentry.o: ../include/ldap.h
-request.o: request.c ../include/lber.h
+addentry.o: ../include/proto-ldap.h
+request.o: request.c ../pgpsockets/pgpSocketsLDAP.h
+request.o: ../../../pgpcdk/pub/include/pgpSockets.h
 request.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 request.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 request.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-request.o: ../../../pgpcdk/pub/include/pgpTLS.h
 request.o: ../../../pgpcdk/pub/include/pgpErrors.h
 request.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
 request.o: ../include/ldap.h
@@ -365,18 +342,14 @@
 os-ip.o: ../../../pgpcdk/pub/include/pgpOptionList.h
 os-ip.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 os-ip.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-os-ip.o: ../../../pgpcdk/pub/include/pgpTLS.h
 os-ip.o: ../../../pgpcdk/pub/include/pgpErrors.h
-os-ip.o: ../../../pfl/common/pgpPFLErrors.h ../include/proto-ldap.h
-url.o: url.c ../pgpsockets/pgpSocketsLDAP.h
-url.o: ../../../pgpcdk/pub/include/pgpSockets.h
-url.o: ../../../pgpcdk/pub/include/pgpOptionList.h
+os-ip.o: ../../../pfl/common/pgpPFLErrors.h ../include/ldap.h
+url.o: url.c ../include/lber.h ../../../pgpcdk/pub/include/pgpOptionList.h
 url.o: ../../../pgpcdk/pub/include/pgpPubTypes.h
 url.o: ../../../pgpcdk/unix/pgpConfig.h ../../../pfl/common/pgpBase.h
-url.o: ../../../pgpcdk/pub/include/pgpTLS.h
 url.o: ../../../pgpcdk/pub/include/pgpErrors.h
 url.o: ../../../pfl/common/pgpPFLErrors.h ../include/pgpSrvStd.h
-url.o: ../include/ldap.h
+url.o: ../include/proto-ldap.h
 charset.o: charset.c
 
 # IF YOU PUT ANYTHING HERE IT WILL GO AWAY
@@ -459,33 +432,24 @@
 FORCE:
 
 #
-# LDAP SunOS5 GNU C Make-platform file
+# LDAP AIX standard cc Make-platform file
 #
 
+
 #
 # add any platform-specific overrides below here
 #
 
-
 #
 # -------------------------------------------------------------------------
 # you will probably not need to edit anything below this point
 # -------------------------------------------------------------------------
 CC	= gcc
-
-# ranlib not needed under SunOS5
-RANLIB	= ""
-
-# give full path to hostname since it may not be in user's path
-HOSTNAME=/usr/ucb/hostname
 
-# don't count on /usr/ucb/install being present or first in path
-INSTALL=$(LDAPSRC)/build/install.sh
+# install with BSD commandline
+INSTALL=/usr/ucb/install
 
-PLATFORMCFLAGS=	-Dsunos5 -D_REENTRANT
-PLATFORMLIBS= -lsocket -lnsl -lgen
-THREADS= -DTHREAD_SUNOS5_LWP
-THREADSLIB=-lthread
+PLATFORMCFLAGS=	-Daix
 #-----------------------------------------------------------------------------
 # Copyright (c) 1997 PGP inc.
 # All rights reserved.
Index: libs/pfl/configure.in
===================================================================
RCS file: /pgpsrc/libs/pfl/configure.in,v
retrieving revision 1.12
retrieving revision 1.12.2.5.2.3
diff -u -r1.12 -r1.12.2.5.2.3
--- configure.in	1999/04/19 17:59:53	1.12
+++ configure.in	1999/07/20 21:11:03	1.12.2.5.2.3
@@ -27,17 +27,23 @@
 *-*-solaris*)
     PGPARCH=PGP_UNIX_SOLARIS
     ;;
-*-*-linux)
+*-*-linux*)
     PGPARCH=PGP_UNIX_LINUX
     ;;
 *-*-hpux*)
     PGPARCH=PGP_UNIX_HPUX
     ;;
+*-*-aix*)
+    PGPARCH=PGP_UNIX_AIX
+    ;;
 *)
     PGPARCH=PGP_UNIX_UNKNOWN
     ;;
 esac
 
+dnl Need this line if we ever want to use PGPARCH
+dnl AC_SUBST(PGPARCH)
+
 dnl Checks for programs.
 # Find a compiler to use.
 # Check 1) The $CC environment varaible, 2) gcc, 3) acc, and 4) cc.
@@ -149,13 +155,21 @@
 
 # Prefer optimizing, with debugging a second
 if test $ac_cv_prog_gcc$ac_cv_prog_gcc_g = yesyes; then
-  CFLAGS="-O -g"
+  # If we're compiling on AIX, we can't include debug info.  If
+  # we do, we get evil linker errors.
+  # There can be problems with "TOC overflow" on some versions of
+  # AIX, and "-mminimal-toc" is a suggested workaround.
+  if test $PGPARCH = PGP_UNIX_AIX; then
+    CFLAGS="-mminimal-toc -O"
+  else
+    CFLAGS="-O -g"
+  fi
 elif test $ac_cv_prog_suncc = yes; then
   CFLAGS="-O -g"
 elif test $ac_cv_prog_hpuxcc = yes; then
   CFLAGS="-O -Dinline= -Ae"
 else
-  CFLAGS=-O
+  CFLAGS="-O"
 fi
 
 dnl This is to get warnings enabled where possible.
@@ -228,8 +242,12 @@
 AC_SUBST(TEST_L)
 
 dnl Checks for libraries.
-AC_CHECK_LIB(thread, thr_create)
-AC_CHECK_LIB(pthread, pthread_create)
+AC_CHECK_LIB(pthreads, pthread_mutex_init, THREADLIB="-lpthreads",
+  AC_CHECK_LIB(pthread, pthread_mutex_init, THREADLIB="-lpthread",
+    AC_CHECK_LIB(thread, thread_create, THREADLIB="-lthread")))
+
+
+
 
 dnl Checks for library functions.
 AC_CHECK_FUNCS(gethrtime clock_gettime clock_getres gettimeofday \
Index: libs/pfl/common/pgpBase.h
===================================================================
RCS file: /pgpsrc/libs/pfl/common/pgpBase.h,v
retrieving revision 1.27
retrieving revision 1.27.2.1.2.2
diff -u -r1.27 -r1.27.2.1.2.2
--- pgpBase.h	1999/04/19 17:59:53	1.27
+++ pgpBase.h	1999/07/20 21:11:04	1.27.2.1.2.2
@@ -7,13 +7,18 @@
 	This file deals with system dependencies to derive our very basic data
 	types.  It should not contain any higher level types.
 
-	$Id: pgpBase.h,v 1.27 1999/04/19 17:59:53 melkins Exp $
+	$Id: pgpBase.h,v 1.27.2.1.2.2 1999/07/20 21:11:04 sluu Exp $
 ____________________________________________________________________________*/
 #ifndef Included_pgpBase_h	/* [ */
 #define Included_pgpBase_h
 
 #include "pgpPFLConfig.h"
 
+
+#include <stddef.h>
+
+
+
 #if !( defined(PGP_MACINTOSH) || defined(PGP_UNIX) || defined(PGP_WIN32) )
 #error one of {PGP_MACINTOSH, PGP_UNIX, PGP_WIN32} must be defined
 #endif
@@ -189,7 +194,14 @@
 typedef void *				PGPUserValue;
 
 /* A PGPSize refers to in memory sizes. Use PGPFileOffset for file offsets */
+#if PGP_UNIX_HPUX
+/* HPUX has conflicting types for size_t.  This forces a PGPSize to 
+ * always be the same type. */
+typedef unsigned long       PGPSize;
+#else
 typedef size_t				PGPSize;
+#endif
+
 #define MAX_PGPSize			( ~(PGPSize)0 )
 
 /* An offset or size of a file */
Index: libs/pfl/common/pgpMemoryMgrStd.c
===================================================================
RCS file: /pgpsrc/libs/pfl/common/pgpMemoryMgrStd.c,v
retrieving revision 1.7
retrieving revision 1.7.4.1
diff -u -r1.7 -r1.7.4.1
--- pgpMemoryMgrStd.c	1999/03/10 02:53:24	1.7
+++ pgpMemoryMgrStd.c	1999/06/10 04:12:46	1.7.4.1
@@ -6,7 +6,7 @@
 		pgpCreateStandardMemoryMgr
 		pgpDisposeStandardMemoryMgrUserValue
 		
-	$Id: pgpMemoryMgrStd.c,v 1.7 1999/03/10 02:53:24 heller Exp $
+	$Id: pgpMemoryMgrStd.c,v 1.7.4.1 1999/06/10 04:12:46 dallen Exp $
 ____________________________________________________________________________*/
 #include "pgpPFLErrors.h"
 #include "pgpMem.h"
@@ -151,7 +151,12 @@
 	
 	if ( nonPageable )
 	{
+#if !( defined( __osf__ ) || defined( _AIX ) || defined( __hpux ) || \
+defined( _M_XENIX ) )
+
 		(void)munlock( mem,  numBytes );
+
+#endif /* !( __osf__ || _AIX || __hpux || _M_XENIX ) */
 	}
 }
 
Index: libs/pfl/common/classes/StPGPRefs.h
===================================================================
RCS file: /pgpsrc/libs/pfl/common/classes/StPGPRefs.h,v
retrieving revision 1.13.6.3
retrieving revision 1.13.2.3
diff -u -r1.13.6.3 -r1.13.2.3
--- StPGPRefs.h	1999/06/04 19:00:52	1.13.6.3
+++ StPGPRefs.h	1999/06/24 23:12:34	1.13.2.3
@@ -4,7 +4,7 @@
 	
 	
 	
-	$Id: StPGPRefs.h,v 1.13.6.3 1999/06/04 19:00:52 heller Exp $
+	$Id: StPGPRefs.h,v 1.13.2.3 1999/06/24 23:12:34 bwilcox Exp $
 ____________________________________________________________________________*/
 
 #ifndef Included_StPGPRefs_h	/* [ */
@@ -15,7 +15,7 @@
 class StPGPRef {
 public:
 								StPGPRef()
-									{ Set((T) NULL); }
+									{ Set(static_cast<T>(NULL)); }
 								StPGPRef(const T & inRef)
 									{ Set(inRef); }
 #if PGP_COMPILER_SUN
@@ -30,7 +30,7 @@
 	T							Get()
 									{ return mRef; }
 	void						Free()
-									{ if (mRef != (T) NULL) Dispose(); mRef = (T) NULL; }
+									{ if (mRef != static_cast<T>(NULL)) Dispose(); mRef = static_cast<T>(NULL); }
 						
 								operator T() { return mRef; }
 	T*							operator &() { return &mRef; }
@@ -68,6 +68,7 @@
 #endif
 
 #ifdef Included_pgpMemoryMgr_h
+	inline void StPGPRef<PGPByte *>::Dispose() { PGPFreeData(mRef); }
 	class StPGPDataRef	:	public StPGPRef<PGPByte *> {
 	public:
 						StPGPDataRef() { }
@@ -85,7 +86,6 @@
 #endif
 						operator void*() { return mRef; }
 	};
-	inline void 	StPGPRef<PGPByte *>::Dispose() { PGPFreeData(mRef); }
 
 	typedef StPGPRef<PGPMemoryMgrRef>	StPGPMemoryMgrRef;
 	inline void StPGPMemoryMgrRef::Dispose() { PGPFreeMemoryMgr(mRef); }
Index: libs/pfl/unix/pgpPFLConfig.h.in
===================================================================
RCS file: /pgpsrc/libs/pfl/unix/pgpPFLConfig.h.in,v
retrieving revision 1.9
retrieving revision 1.9.2.1
diff -u -r1.9 -r1.9.2.1
--- pgpPFLConfig.h.in	1999/03/25 18:15:54	1.9
+++ pgpPFLConfig.h.in	1999/06/10 04:08:39	1.9.2.1
@@ -3,7 +3,7 @@
  * the configuration information for the PFL, and it should be
  * included in all PFL source files.
  *
- * $Id: pgpPFLConfig.h.in,v 1.9 1999/03/25 18:15:54 melkins Exp $
+ * $Id: pgpPFLConfig.h.in,v 1.9.2.1 1999/06/10 04:08:39 dallen Exp $
  */
 
 #ifndef Included_pgpPFLConfig_h	/* [ */
@@ -78,7 +78,7 @@
  * Define "PGP_UNIX" if we are on PGP_UNIX and "PGP_UNIX" is
  * not already defined.
  */
-#if defined(unix) || defined(__unix) || defined (__unix__)
+#if defined(unix) || defined(__unix) || defined (__unix__) || defined (_AIX)
 #ifndef PGP_UNIX
 #define PGP_UNIX 1
 #endif
Index: libs/pgpcdk/configure.in
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/configure.in,v
retrieving revision 1.21
retrieving revision 1.21.2.7.2.2
diff -u -r1.21 -r1.21.2.7.2.2
--- configure.in	1999/04/29 20:02:27	1.21
+++ configure.in	1999/07/20 21:11:04	1.21.2.7.2.2
@@ -29,16 +27,30 @@
 case "$host" in
 *-*-solaris*)
     PGPARCH=PGP_UNIX_SOLARIS
+    OSTYPE=solaris
     ;;
-*-*-linux)
+*-*-linux*)
     PGPARCH=PGP_UNIX_LINUX
+    OSTYPE=linux
     ;;
+*-*-aix*)
+    PGPARCH=PGP_UNIX_AIX
+    OSTYPE=aix
+    ;;
+*-*-hpux*)
+    PGPARCH=PGP_UNIX_HPUX
+    OSTYPE=hpux
+    ;;
 *)
     PGPARCH=PGP_UNIX_UNKNOWN
+    OSTYPE=unknown
     ;;
 esac
 
+AC_SUBST(PGPARCH)
+AC_SUBST(OSTYPE)
 
+
 # Export the host CPU to Makefiles for assembly-language use
 HOST_CPU=$host_cpu
 AC_SUBST(HOST_CPU)
@@ -174,13 +186,28 @@
 
 # Prefer optimizing, with debugging a second
 if test $ac_cv_prog_gcc$ac_cv_prog_gcc_g = yesyes; then
-  CFLAGS="-O -g"
+  # If we're compiling on AIX, we can't include debug info.  If
+  # we do, we get evil linker errors.
+  # There can be problems with "TOC overflow" on some versions of
+  # AIX, and "-mminimal-toc" is a suggested workaround.
+  if test $PGPARCH = PGP_UNIX_AIX; then
+    CFLAGS="-mminimal-toc -O"
+  else
+    CFLAGS="-O -g"
+  fi
 elif test $ac_cv_prog_suncc = yes; then
   CFLAGS="-O -g"
 else
   CFLAGS=-O
+fi
+
+# gcc 2.8.1 on HPUX has a bug when optimization is set to anything
+# BUT -O0, so we can't optimize on HPUX.
+if test $PGPARCH = PGP_UNIX_HPUX; then
+  CFLAGS="-O0"
 fi
- 
+
+
 dnl This is to get warnings enabled where possible.
 AC_MSG_CHECKING(for useful warning options (\$WARN))
 # This currently only deals with gcc and Sun's V4.2 compiler.
@@ -263,6 +290,7 @@
 elif test "$bn_tune_guess" = yes; then
   AC_MSG_WARN([architecture guessed.  If incorrect, use explicit \$TUNE.])
 fi
+
 AC_SUBST(TUNE)
 
 fi
@@ -341,9 +369,15 @@
 AC_SUBST(TEST_L)
 
 dnl Checks for libraries.
-AC_CHECK_LIB(thread, main, 
-  THREADLIB="-lthread", 
-  AC_CHECK_LIB(pthread, main, THREADLIB="-lpthread"))
+AC_CHECK_LIB(pthreads, pthread_mutex_init, THREADLIB="-lpthreads",
+  AC_CHECK_LIB(pthread, pthread_mutex_init, THREADLIB="-lpthread",
+    AC_CHECK_LIB(thread, thread_create, THREADLIB="-lthread")))
+
+
+# //!! Note: I switched these to make pthreads be detected 
+# first and libthread to be detected only in the case that 
+# pthreads were not present.  --Bryce 1999/06/156
+
 AC_SUBST(THREADLIB)
 
 dnl Checks for library functions.
Index: libs/pgpcdk/priv/crypto/hash/pgpHash.c
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/priv/crypto/hash/pgpHash.c,v
retrieving revision 1.40
retrieving revision 1.40.18.2
diff -u -r1.40 -r1.40.18.2
--- pgpHash.c	1999/03/10 02:51:17	1.40
+++ pgpHash.c	1999/07/20 21:11:05	1.40.18.2
@@ -4,7 +4,7 @@
 	Copyright (C) 1997 Network Associates Inc. and affiliated companies.
 	All rights reserved.
 
-	$Id: pgpHash.c,v 1.40 1999/03/10 02:51:17 heller Exp $
+	$Id: pgpHash.c,v 1.40.18.2 1999/07/20 21:11:05 sluu Exp $
 ____________________________________________________________________________*/
 #include "pgpConfig.h"
 #include "pgpErrors.h"
@@ -361,7 +361,7 @@
  * Given a hash name, return the corresponding hash.
  */
 	PGPHashVTBL const *
-pgpHashByName (char const *name, size_t namelen)
+pgpHashByName (char const *name, PGPSize namelen)
 {
 	PGPUInt32	algIndex;
 
Index: libs/pgpcdk/priv/crypto/pipe/file/pgpPrsAsc.c
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/priv/crypto/pipe/file/pgpPrsAsc.c,v
retrieving revision 1.50.6.1
retrieving revision 1.50.2.3.2.2
diff -u -r1.50.6.1 -r1.50.2.3.2.2
--- pgpPrsAsc.c	1999/06/03 23:26:00	1.50.6.1
+++ pgpPrsAsc.c	1999/07/20 21:11:05	1.50.2.3.2.2
@@ -9,7 +9,7 @@
  *
  * Written by:	Derek Atkins <warlord@MIT.EDU>
  *
- * $Id: pgpPrsAsc.c,v 1.50.6.1 1999/06/03 23:26:00 heller Exp $
+ * $Id: pgpPrsAsc.c,v 1.50.2.3.2.2 1999/07/20 21:11:05 sluu Exp $
  */
 
 #include "pgpConfig.h"
@@ -232,7 +232,7 @@
 } PrsAscContext;
 
 
-static char const *	sDearmorTable = NULL;	/* enforce 'const' access */
+static signed char const *	sDearmorTable = NULL;	/* enforce 'const' access */
 
 /* forward references */
 static PGPError writeMessage (PrsAscContext *ctx, Message *msg);
@@ -1138,11 +1138,11 @@
 pgpParseAscInit (void)
 {
 	#define kDearmorTableSize		( (PGPSize)256 )
-	static char 		sDearmorTableStorage[ kDearmorTableSize ];
+	static signed char 		sDearmorTableStorage[ kDearmorTableSize ];
 	static PGPBoolean	sInited = FALSE;
 	
 	int		i;
-	char	tempTable[ kDearmorTableSize ];
+	signed char	tempTable[ kDearmorTableSize ];
 
 	if ( sInited )
 		return;
@@ -1151,7 +1151,7 @@
 		tempTable[i] = -1;
 
 	for (i = 0; i < (int)(sizeof(armorTable) - 1); i++)
-		tempTable[armorTable[i] & 0xff] = (char)i;
+		tempTable[armorTable[i] & 0xff] = (signed char)i;
 
 	pgpCopyMemory( tempTable, sDearmorTableStorage, sizeof(tempTable) );
 	sDearmorTable	= &sDearmorTableStorage[ 0 ];
@@ -1281,28 +1281,18 @@
 	while (inlen && isspace (in[inlen - 1]))
 		inlen--;
 
-	if (*in != '=')
-		return -1;
-	if (inlen == 5)
-	{
-		if (dearmorMorsel (in + 1, buf) != 4)
-		{
-			return -1;
-		}
-		else if (inlen == 7)
-		{
-			if (in[1] != '3' || in[2] != 'D' ||
-			    dearmorMorsel (in + 3, buf) != 4)
-			{
-				return -1;
-			}
-			else
-			{
-				return -1;
-			}
-		}
-	}
-	
+        if (*in != '=')
+                return -1;
+        if (inlen == 5)
+                if (dearmorMorsel (in + 1, buf) != 4)
+                        return -1;
+                else if (inlen == 7)
+                        if (in[1] != '3' || in[2] != 'D' ||
+                            dearmorMorsel (in + 3, buf) != 4)
+                                return -1;
+                        else
+                                return -1;
+
 	/*
 	 * Welcome to the famous ANSI C glitch.  ANSI C
 	 * promotes to signed values where possible when preserving
Index: libs/pgpcdk/priv/crypto/random/pgpRndUnix.c
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/priv/crypto/random/pgpRndUnix.c,v
retrieving revision 1.13.6.1
retrieving revision 1.13.2.2
diff -u -r1.13.6.1 -r1.13.2.2
--- pgpRndUnix.c	1999/06/04 00:28:48	1.13.6.1
+++ pgpRndUnix.c	1999/06/24 23:12:44	1.13.2.2
@@ -2,7 +2,7 @@
  * Get high-resolution timing information to seed the random number generator.
  * Unix version.  (The complicated one.)
  *
- * $Id: pgpRndUnix.c,v 1.13.6.1 1999/06/04 00:28:48 heller Exp $
+ * $Id: pgpRndUnix.c,v 1.13.2.2 1999/06/24 23:12:44 bwilcox Exp $
  */
 
 #include "pgpConfig.h"
@@ -147,12 +147,12 @@
 #ifndef USE_FTIME
 #ifndef HAVE_FTIME
 #define HAVE_FTIME 1
-#endif
+#endif /* #ifndef HAVE_FTIME */
 #ifndef HAVE_SYS_TIMEB_H
 #define HAVE_SYS_TIMEB_H 1
-#endif
+#endif /* #ifndef HAVE_SYS_TIMEB_H */
 #define USE_FTIME HAVE_FTIME && HAVE_SYS_TIMEB_H
-#endif
+#endif /* #ifndef USE_FTIME */
 
 #if USE_FTIME
 #include <sys/timeb.h>
Index: libs/pgpcdk/priv/external/common/tis/cms/src/cert_struct.h
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/priv/external/common/tis/cms/src/cert_struct.h,v
retrieving revision 1.6.14.1
retrieving revision 1.6
diff -u -r1.6.14.1 -r1.6
--- cert_struct.h	1999/07/09 21:53:14	1.6.14.1
+++ cert_struct.h	1999/05/19 01:05:26	1.6
@@ -547,5 +547,4 @@
 } PKICRLDistPointsSyntax;
 #define PKIID_CRLDistPointsSyntax PKIID_SEQUENCE_OF
 
-
 #endif /* cert_STRUCT_H_ */
Index: libs/pgpcdk/priv/networklib/pgpsockets/common/pgpSockets.c
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/priv/networklib/pgpsockets/common/pgpSockets.c,v
retrieving revision 1.25.6.2
retrieving revision 1.25.2.2
diff -u -r1.25.6.2 -r1.25.2.2
--- pgpSockets.c	1999/06/04 23:33:17	1.25.6.2
+++ pgpSockets.c	1999/06/24 23:12:54	1.25.2.2
@@ -4,7 +4,7 @@
 	
 	
 	
-	$Id: pgpSockets.c,v 1.25.6.2 1999/06/04 23:33:17 heller Exp $
+	$Id: pgpSockets.c,v 1.25.2.2 1999/06/24 23:12:54 bwilcox Exp $
 ____________________________________________________________________________*/
 
 #if PGP_WIN32
@@ -39,7 +39,7 @@
 # define closesocket(x) close((x))
 # define ioctlsocket(a,b,c) ioctl((a),(b),(c))
 typedef int		SOCKET;
-#if PGP_UNIX_LINUX || PGP_UNIX_HPUX
+#if PGP_UNIX_LINUX || PGP_UNIX_HPUX || PGP_UNIX_AIX
 int gethostname (char * name, size_t namelen);
 #else
 int gethostname (char * name, int namelen);
Index: libs/pgpcdk/pub/include/pgpSockets.h
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/pub/include/pgpSockets.h,v
retrieving revision 1.27.6.1
retrieving revision 1.27.2.3
diff -u -r1.27.6.1 -r1.27.2.3
--- pgpSockets.h	1999/06/04 23:33:17	1.27.6.1
+++ pgpSockets.h	1999/06/25 01:52:04	1.27.2.3
@@ -4,7 +4,7 @@
 	
 	
 	
-	$Id: pgpSockets.h,v 1.27.6.1 1999/06/04 23:33:17 heller Exp $
+	$Id: pgpSockets.h,v 1.27.2.3 1999/06/25 01:52:04 bwilcox Exp $
 ____________________________________________________________________________*/
 
 
@@ -25,6 +25,9 @@
 # include <sys/ioctl.h>		/* Need FIONREAD */
 #elif PGP_UNIX_SOLARIS
 # include <sys/filio.h>
+#elif PGP_UNIX_AIX
+# include <sys/time.h>
+# include <sys/ioctl.h>
 #endif /* ! PGP_UNIX_LINUX */
 # include <netinet/in.h>
 # include <netdb.h>
@@ -65,7 +68,7 @@
  typedef struct sockaddr_in	PGPSocketAddressInternet;
  typedef struct sockaddr	PGPSocketAddress;
  typedef struct in_addr		PGPInternetAddress;
- typedef fd_set 			PGPSocketSet;
+ typedef fd_set 		PGPSocketSet;
  typedef struct timeval		PGPSocketsTimeValue;
 
 # define PGPSOCKETSET_CLEAR(socketRef, set) FD_CLR((int) (socketRef), (set))
Index: libs/pgpcdk/unix/config/pre.in
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/unix/config/pre.in,v
retrieving revision 1.18.10.1
retrieving revision 1.18.14.2
diff -u -r1.18.10.1 -r1.18.14.2
--- pre.in	1999/07/01 01:47:39	1.18.10.1
+++ pre.in	1999/07/20 21:11:12	1.18.14.2
@@ -4,7 +4,7 @@
 # Copyright (C) 1996,1997 Network Associates Inc. and affiliated companies.
 # All rights reserved.
 #
-# $Id: pre.in,v 1.18.10.1 1999/07/01 01:47:39 mmonroe Exp $
+# $Id: pre.in,v 1.18.14.2 1999/07/20 21:11:12 sluu Exp $
 #
 
 SHELL=/bin/sh
@@ -27,7 +27,7 @@
 pubinc=$(top_builddir)/pub/include
 pubincback=../..
 privinc=$(top_builddir)/priv/include
-bsafeinc=$(top_builddir)/priv/external/unix/bsafe/include
+bsafeinc=$(top_builddir)/priv/external/unix/bsafe/include/@OSTYPE@
 rsarefinc=$(top_builddir)/priv/external/common/rsaref/source
 privincback=../..
 opaqueinc=$(top_builddir)/priv/include/opaque
Index: libs/pgpcdk/unix/ui/PGPPassphraseDialog.cpp
===================================================================
RCS file: /pgpsrc/libs/pgpcdk/unix/ui/PGPPassphraseDialog.cpp,v
retrieving revision 1.7
retrieving revision 1.7.2.1
diff -u -r1.7 -r1.7.2.1
--- PGPPassphraseDialog.cpp	1999/04/19 17:59:55	1.7
+++ PGPPassphraseDialog.cpp	1999/06/10 04:31:31	1.7.2.1
@@ -2,7 +2,7 @@
 	Copyright (C) 1997 Network Associates Inc. and affiliated companies.
 	All rights reserved.
 	
-	$Id: PGPPassphraseDialog.cpp,v 1.7 1999/04/19 17:59:55 melkins Exp $
+	$Id: PGPPassphraseDialog.cpp,v 1.7.2.1 1999/06/10 04:31:31 dallen Exp $
 ____________________________________________________________________________*/
 
 #include <stdio.h>
@@ -166,7 +166,7 @@
 
 		// get name on key
 		PGPGetPrimaryUserIDNameBuffer (Key, sizeof(szName), 
-										szName, &u);
+										szName, (PGPSize *) &u);
 		if (u > MAXDECRYPTIONNAMECHAR) 
 		{
 			u = MAXDECRYPTIONNAMECHAR;
@@ -276,7 +276,7 @@
 	strcat (sz2, ")");
 
 	// get name on key
-	PGPGetPrimaryUserIDNameBuffer (Key, sizeof(szName),szName, &u);
+	PGPGetPrimaryUserIDNameBuffer (Key, sizeof(szName),szName, (PGPSize *) &u);
 
 	strcpy(szNameFinal, szName);
 	strcat(szNameFinal, sz2);
