*** /usr/src/sys/sys/malloc.h	Fri Feb  3 03:39:48 1995
--- malloc.h	Thu Sep 28 15:02:44 1995
***************
*** 1,3 ****
--- 1,12 ----
+ /*
+  * This file has been slightly modified by NRL for use with IPv6.  Search
+  * for 'INET6' to see the ifdef blocks where this happens.
+  * (Actually, the change here was to add more malloc types.  Search for
+  *  'IPv6'.)
+  *
+  * See the NRL Copyright notice for conditions on the modifications.
+  */
+ 
  /*	BSDI malloc.h,v 2.1 1995/02/03 08:39:48 polk Exp	*/
  
  /*
***************
*** 113,118 ****
--- 122,132 ----
  #define	M_MSDOSFSMNT	61	/* MSDOSFS mount structure */
  #define	M_MSDOSFSFAT	62	/* MSDOSFS FAT table */
  #define	M_MSDOSFSNODE	63	/* MSDOSFS node structure */
+ 
+ #define M_DISCQ         64      /* IPv6 discovery queue */
+ #define M_FRAGQ         65      /* IPv6 fragment queue */
+ #define M_SECA          66      /* IPsec Security Association queue */
+ 
  #define	M_TEMP		74	/* misc temporary data buffers */
  #define	M_LAST		75	/* Must be last type + 1 */
  
***************
*** 181,188 ****
  	"MSDOSFS mount",/* 61 M_MSDOSFSMNT */ \
  	"MSDOSFS fat",	/* 62 M_MSDOSFSFAT */ \
  	"MSDOSFS node", /* 63 M_MSDOSFSNODE */ \
! 	NULL, NULL, NULL, NULL, NULL, \
! 	NULL, NULL, NULL, NULL, NULL, \
  	"temp",		/* 74 M_TEMP */ \
  }
  
--- 195,205 ----
  	"MSDOSFS mount",/* 61 M_MSDOSFSMNT */ \
  	"MSDOSFS fat",	/* 62 M_MSDOSFSFAT */ \
  	"MSDOSFS node", /* 63 M_MSDOSFSNODE */ \
!         "IPv6 discq",   /* 64 M_DISCQ */ \
!         "IPv6 fragq",   /* 65 M_FRAGQ */ \
!         "Sec Assoc",    /* 66 M_SECA */ \
! 	NULL, NULL, NULL, NULL, \
! 	NULL, NULL, NULL, \
  	"temp",		/* 74 M_TEMP */ \
  }
  
